BUG: Fixes #23644: sparse: csc_array.setdiag() makes indices unsorted

broesler opened the pull request:
#### Reference issue
Fixes #23644,

#### What does this implement/fix?
This PR:

* fixes #23644 by sorting COO indices in `_cs_matrix._setdiag`.
* refactors `mask` -> `is_existing` and `is_new` in `_setdiag`.
* fixes the condition for converting to COO format in `_setdiag`.
* adds tests for the logic branches in `_setdiag`.

This PR checks whether `has_sorted_indices` is set in the COO branch, and sorts the
newly-created diagonal indices accordingly.

It also updates the condition for converting to COO format in `_setdiag`.
The previous condition was:

```python
self.nnz - mask.sum() < self.nnz * 0.001
```

where `mask` was the logical array that was True for existing elements
on the `k`-th diagonal. This condition did not have the intended effect
when there were non-zeros *not* on the diagonal being set. The new
condition is:

```python
N_new < self.nnz * 0.001
```

where `N_new` is the number of non-existing elements on the `k`-th
diagonal.

This PR also refactors the `mask` array to be named explicitly
`is_existing` because it represents the logical indexing of existing
elements. Similarly, `is_new = ~is_existing`.

This PR creates a new mixin class for testing the
`_cs_matrix._setdiag` function branches from gh-19962. 

The existing `test_setdiag_cs[cr]` methods are now encompassed by the
`_CompressedMixin.test_setdiag_noconvert` method. Remove them to avoid
repeating logic and redundant tests.

#### Additional information
Just to note, the `_cs_matrix._setdiag` function, and similarly the `_coo_base._setdiag` function, cannot create duplicate
entries, so we don't actually need to check `has_canonical_format` or run
`sum_duplicates()`. Both `_setdiag` functions allow explicit zeros,
but that is not part of the SciPy definition of "canonical" format. Thus, we can just check `has_sorted_indices` and run
`sort_indices()` if necessary.

broesler committed:
TST: Create _CompressedMixin class for test_setdiag* tests.

This test creates a new mixin class for testing the
`_cs_matrix._setdiag` function branches from gh-19962. As-is, the
`_setdiag` function has a bug in the convert to COO branch that does not
sort the indices if `self.has_sorted_indices` is set. The test is marked
with an `xfail`.

broesler committed:
TST: Remove test_setdiag_cs[cr].

The existing `test_setdiag_cs[cr]` methods are now encompassed by the
`_CompressedMixin.test_setdiag_noconvert` method. Remove them to avoid
repeating logic and redundant tests.

broesler committed:
BUG/MNT: Update condition for converting to COO format in _setdiag.

This commit changes the condition for converting from CSC/CSR format to
COO format in `_cs_matrix._setdiag.` The previous condition was:

  self.nnz - mask.sum() < self.nnz * 0.001

where `mask` was the logical array that was True for existing elements
on the k-th diagonal. This condition did not have the intended effect
when there were non-zeros *not* on the diagonal being set. The new
condition is:

  N_new < self.nnz * 0.001

where `N_new` is the number of non-existing elements on the k-th
diagonal.

This commit refactors the `mask` array to be named explicitly
`is_existing` because it represents the logical indexing of existing
elements. Similarly, `is_new = ~is_existing`.

broesler committed:
BUG: Fixes gh-23644 by sorting COO indices in _cs_matrix._setdiag.

This commit checks whether `has_sorted_indices` is set, and sorts the
newly-created diagonal indices accordingly.

broesler committed:
TST: Remove xfail from test_setdiag_cooconvert.

The test now passes since the bug is fixed, so remove the xfail.

dschult reviewed (commented):
Thanks @broesler for this!
I have a comment below about when to check for sorted indices. I'm in favor of matching what `insert_many` does with `do_sort`.

Also there is a bigger issue here about whether the sort should be done after converting back to `cs[rc]`, or whether it should be done before converting back (that is, while still in COO format).  I'm not sure which is better. So maybe that answer will wait until more testing has been done. 

We have to reevaluate the timing to check the formula with this new behavior anyway. So I think its OK to put off checking where the sort should occur to another PR where we re-examine the formula.

broesler committed:
BUG: Fix potential bug in evaluating has_sorted_indices in _setdiag.

This commit is in response to a reviewer comment, noting that
`has_sorted_indices` should be cached before changes are made to the
matrix, like in `_insert_many`.

broesler commented on scipy/sparse/_compressed.py:
```diff
@@ -804,23 +804,27 @@ def _setdiag(self, values, k):
             arrays = coo._coo_to_compressed(self._swap)
             self.indptr, self.indices, self.data, _ = arrays
+            # Sort the indices (like in _insert_many)
+            if self.has_sorted_indices:
```
Thanks for pointing out this issue. I moved the `do_sort` line before the `self.tocoo()` etc.

broesler committed:
ENH: Move is_new allocation to relevant branch.

This commit is in response to a reviewer suggestion to move the
allocation of `is_new` in `_cs_matrix._setdiag` to the relevant logic
branch. It delays allocation until necessary.

broesler committed:
ENH: Use exsiting memory to compute logical not in _cs_matrix._setdiag.

This commit is in response to a reviewer comment. It uses
alraedy-allocated memory to compute a logical not array and reduce RAM
footprint.

dschult reviewed (approved):
This looks good to me. I think all of @perimosocordiae suggestions are included as well.

I approve this PR.
